apps/server/lib/lexical/server/code_intelligence/entity.ex

```diff
@@ -0,0 +1,240 @@
+    unsupported_node(node)
+  end
+
+  defp unsupported_node(node) do
+    {:error, {:unsupported, node}}
+  end
+
+  defp fetch_current_module_aliases(nil), do: {:error, :missing_defmodule}
+
+  defp fetch_current_module_aliases(zipper) do
+    case Zipper.node(zipper) do
+      {:defmodule, _, [{:__aliases__, _, aliases} | _]} ->
+        {:ok, Enum.map(aliases, &unwrap_atom/1)}
+
+      _ ->
+        zipper |> Zipper.up() |> fetch_current_module_aliases()
+    end
+  end
+
+  defp unwrap_atom({:__block__, _, [atom]}) when is_atom(atom) do
+    atom
+  end
+
+  defp unwrap_atom(atom) when is_atom(atom) do
+    atom
+  end
+
+  defp innermost_zipper_at(ast, position) do
+    %{line: line, character: char} = position
+
+    {_, {innermost, _}} =
+      ast
+      |> Zipper.zip()
+      |> Zipper.traverse_while({nil, false}, fn
+        {{_, _, _} = node, _} = zipper, {current, found_line?} ->
+          case fetch_range(node) do
+            {:ok, range} ->
+              same_line? = line == range.start[:line] and line == range.end[:line]
+              line_in_range? = line >= range.start[:line] and line <= range.end[:line]
+              # sourceror column ranges are inclusive at the start and exclusive at the end
+              char_in_range? = char >= range.start[:column] and char < range.end[:column]
+
+              cond do
+                same_line? and char_in_range? ->
+                  {:cont, zipper, {zipper, true}}
+
+                found_line? or same_line? ->
+                  {:skip, zipper, {current, true}}
+
+                line_in_range? ->
+                  {:cont, zipper, {zipper, false}}
+
+                true ->
+                  {:skip, zipper, {current, found_line?}}
+              end
+
+            _ ->
+              {:cont, zipper, {current, found_line?}}
+          end
+
+        zipper, acc ->
+          {:cont, zipper, acc}
+      end)
+
+    if innermost do
+      {:ok, innermost}
+    else
+      {:error, :not_found}
+    end
+  end
+
+  # Handle aliases explicitly because Sourceror does not expect the atoms
+  # making up the components of the alias to be wrapped in tuples with
+  # additional metadata.
+  defp fetch_range({:__aliases__, _, [_ | _] = args}) do
+    {_, start_meta, _} = List.first(args)
+    {_, end_meta, [atom]} = List.last(args)
+
+    range = %{
+      start: [line: start_meta[:line], column: start_meta[:column]],
+      end: [line: end_meta[:line], column: end_meta[:column] + String.length(to_string(atom))]
+    }
+
+    {:ok, range}
+  end
+
+  # Corrects dot-call issue
+  defp fetch_range({:., _, [left, atom]}) when is_atom(atom) do
+    with {:ok, range} <- fetch_range(left) do
+      # Unhandled edge case: quoted atoms, e.g. Foo."bar-baz"
+      len = atom |> Atom.to_string() |> String.length()
+      {:ok, update_in(range.end[:column], &(&1 + 1 + len))}
+    end
+  end
+
+  # There are certain AST nodes that fail on Sourceror.get_range/1, like
```

Comment:
The current implementation appears to be overly complex to me, especially the logic for finding the innermost node and fetching the range.

This leads me to question whether it's really necessary to use Sourceror for this task. The power of the zipper lies in its ability to facilitate node modifications, but do we really need to make extensive modifications to nodes? We only need to find the entity under the current cursor, and currently, we only handle modules.

I'm not sure if `cursor_context` + `container_cursor_to_quoted` + `surround_context` can be helpful in this regard. For example, in a case like `AAA.{BBB.C|CC`, perhaps we could first use them to find AAA.C, and then combine it with the remaining `CC`. This approach might simplify the code. We can also move the cursor for other cases if needed.

Reply:
I do not explain the motivation for this strategy in the code or in comments, and that's a mistake. I'll try to do so now, and then, if you all agree that this is a reasonable approach, I'll work on including better explanation in docs/comments such that it's more accessible to others in the future.

First, I'd like to address this comment:

> The power of the zipper lies in its ability to facilitate node modifications...

I don't agree; while zippers certainly do facilitate node modifications, the power of zippers is arbitrary movement. With standard tree traversal, a given node can easily inspect its children, but it cannot inspect its parent or siblings (except with the addition of a complex accumulator, at which point you're reinventing a zipper!).

I'll come back to zippers with a motivating example in a moment.

The ultimate goal of `resolve/2` is to resolve a "code entity" at a specific cursor position. While it currently handles only module aliases (at [@scohen's suggestion](https://github.com/lexical-lsp/lexical/pull/331#issuecomment-1678182334)), it's designed to be extended to many other use-cases. Consider the following AST, using `|` to represent various positions:

```elixir
defmodule |My.Module do # (1)
  alias My.Module.Neste|d # (2)

  @type some_type :: {:ok, Nested.|t()} | :error # (3)

  def fun(%Nested{} = arg) do
    ar|g # (4)
    |> Nested.|process_arg(%|{data: :literal}) # (5) (6)
    |> Nested.finish()
  end
end
```

The entities associated with these positions might be something like:

1. `{:module, My}`
2. `{:module, My.Module.Nested}`
3. `{:type, {My.Module.Nested, t, 0}}`
4. `{:argument, {My.Module, :fun, 1, 0}}`
5. `{:call, {My.Module.Nested, :process_arg, 2}}`
6. `{:literal, :map}`

For `textDocument/hover`, each of these entities might have docs or other useful information associated with them.

For instance, hovering (4) might show the relevant part(s) of the function header that specify that argument. In order to know that this is a bound argument to the function and not a local variable or import, you *must* inspect its context. This is where zipper shine. By exposing the innermost AST node for a position as a zipper, we can now inspect the contextual environment of that node to understand how its being used.

Getting back to modules -- the reason for the custom range stuff and changes to `Ast.parse/` is that, otherwise, this cannot be used for certain multi-line nodes, namely:

```elixir
# this is unusual and not idiomatic, but it's valid
Foo.
  Bar.
  Baz

# by default, this parses into
{:__aliases__, [line: 1, column: 1, last: [line: 3, column: 3], [:Foo, :Bar, :Baz]}
```

If we're hovering over `Bar`, there is no element more specific than the `:__aliases__` node to consider. By using `:static_atoms_encoder`, we get this parse tree, which allows us to consider `Bar` and the alias segments prior:

```elixir
{:__aliases__, [line: 1, column: 1, last: [line: 3, column: 3],
 [{:__block__, [line: 1, column: 1], [:Foo]},
  {:__block__, [line: 2, column: 3], [:Bar]},
  {:__block__, [line: 3, column: 3], [:Baz]}]}
```

In order to use this somewhat modified AST, we have to wrap `Sourceror.get_range`, which doesn't expect these additional `:__block__` nodes.

---

This is getting too long, so to summarize and make a few final points:

1. `resolve/2` is intended as a general-purpose API to discover the *kind of entity* at a position, and that requires understanding its surroundings/context.
2. Zippers are especially well-suited to inspecting the context of a given node and are, I think, appropriate to use here.
3. I think this code could/should be better organized. For instance, `fetch_range` is addressing changes that are happening in `Ast`, and should live/be documented next to that.

Reply:
Disregard; I think this can be built using much of the existing work for completions. I'm going to study that a bit.